File: cpp/src/obs.cpp
#include "hls/obs.hpp"

#include <algorithm>
#include <stdexcept>

namespace hls {

void quat_rotate_inverse(const double q[4], const double v[3], double out[3]) {
  const double w = q[0], x = q[1], y = q[2], z = q[3];
  const double dot = x * v[0] + y * v[1] + z * v[2];
  const double cx = y * v[2] - z * v[1], cy = z * v[0] - x * v[2], cz = x * v[1] - y * v[0];
  const double a = 2 * w * w - 1;
  out[0] = v[0] * a - 2 * w * cx + 2 * x * dot;
  out[1] = v[1] * a - 2 * w * cy + 2 * y * dot;
  out[2] = v[2] * a - 2 * w * cz + 2 * z * dot;
}

void assemble_obs(const Manifest& m, const RobotState& s, const double cmd[3],
                  const float* last_action, float* out) {
  const int n = m.num_actions();
  float* p = out;
  for (const auto& t : m.obs_terms) {
    float* start = p;
    if (t.name == "base_ang_vel") {
      for (double v : s.gyro) *p++ = static_cast<float>(v);
    } else if (t.name == "base_lin_vel") {
      for (double v : s.lin_vel) *p++ = static_cast<float>(v);
    } else if (t.name == "projected_gravity") {
      const double g[3] = {0, 0, -1};
      double r[3];
      quat_rotate_inverse(s.quat.data(), g, r);
      for (double v : r) *p++ = static_cast<float>(v);
    } else if (t.name == "velocity_commands") {
      for (int i = 0; i < 3; ++i) *p++ = static_cast<float>(cmd[i]);
    } else if (t.name == "joint_pos") {
      for (int i = 0; i < n; ++i) *p++ = static_cast<float>(s.q[i] - m.default_joint_pos[i]);
    } else if (t.name == "joint_vel") {
      for (int i = 0; i < n; ++i) *p++ = static_cast<float>(s.qd[i]);
    } else if (t.name == "actions") {
      for (int i = 0; i < n; ++i) *p++ = last_action[i];
    } else {
      throw std::runtime_error("unknown observation term " + t.name);
    }
    if (p - start != t.dim) throw std::runtime_error("obs term size mismatch: " + t.name);
    for (float* v = start; v != p; ++v) {
      *v *= t.scale;
      if (t.clip) *v = std::clamp(*v, t.clip->first, t.clip->second);
    }
  }
}

}  // namespace hls

File: src/humanoid_loco/obs.py
"""Observation terms as pure numpy, mirroring ``isaaclab.envs.mdp`` semantics.

Each term is a function of the robot state, the velocity command and the previous action.
The manifest names which terms appear and in what order; ``assemble_obs`` does the rest.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass

import numpy as np

from humanoid_loco.manifest import PolicyManifest


@dataclass
class RobotState:
    """Minimal proprioception. Joint arrays are in manifest (policy) order."""

    base_quat: np.ndarray  # (4,) w, x, y, z  world->base
    base_ang_vel: np.ndarray  # (3,) rad/s, base frame (IMU gyro)
    joint_pos: np.ndarray  # (n,)
    joint_vel: np.ndarray  # (n,)
    base_lin_vel: np.ndarray | None = None  # (3,) base frame; privileged, critic-only


def quat_rotate_inverse(q: np.ndarray, v: np.ndarray) -> np.ndarray:
    """Rotate world-frame ``v`` into the frame described by unit quaternion ``q`` (w,x,y,z)."""
    w, xyz = q[0], q[1:]
    return v * (2.0 * w * w - 1.0) - 2.0 * w * np.cross(xyz, v) + 2.0 * xyz * np.dot(xyz, v)


def projected_gravity(base_quat: np.ndarray) -> np.ndarray:
    return quat_rotate_inverse(base_quat, np.array([0.0, 0.0, -1.0]))


ObsFn = Callable[[RobotState, np.ndarray, np.ndarray, PolicyManifest], np.ndarray]

TERMS: dict[str, ObsFn] = {
    "base_lin_vel": lambda s, c, a, m: s.base_lin_vel,
    "base_ang_vel": lambda s, c, a, m: s.base_ang_vel,
    "projected_gravity": lambda s, c, a, m: projected_gravity(s.base_quat),
    "velocity_commands": lambda s, c, a, m: c,
    "joint_pos": lambda s, c, a, m: s.joint_pos - np.asarray(m.default_joint_pos),
    "joint_vel": lambda s, c, a, m: s.joint_vel,
    "actions": lambda s, c, a, m: a,
}


def assemble_obs(
    manifest: PolicyManifest,
    state: RobotState,
    command: np.ndarray,
    last_action: np.ndarray,
    group: list | None = None,
) -> np.ndarray:
    """Concatenate manifest terms (or an explicit term list) with their scale and clip."""
    parts = []
    for term in group if group is not None else manifest.obs_terms:
        try:
            value = np.asarray(TERMS[term.name](state, command, last_action, manifest), np.float32)
        except KeyError as e:
            raise KeyError(f"unknown observation term {term.name!r}") from e
        if value.size != term.dim:
            raise ValueError(f"{term.name}: got {value.size} values, manifest says {term.dim}")
        value = value.ravel() * term.scale
        if term.clip is not None:
            value = np.clip(value, *term.clip)
        parts.append(value)
    return np.concatenate(parts)
